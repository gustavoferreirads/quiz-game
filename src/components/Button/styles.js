import { StyleSheet } from 'react-native';
import { colors, fonts } from '../../styles';

const styles = StyleSheet.create({
  container: {
    height: 40,
    borderRadius: 30,
    justifyContent: 'center',
    alignItems: 'center',
    paddingHorizontal: 20,
    fontWeight: 'bold',
    width: 150,
  },

  'button-inline': {
    backgroundColor: colors.primary,
  },

  'button-outline': {
    backgroundColor: '#FFF',
    borderWidth: 1,
    borderColor: colors.primary,
  },

  text: {
    fontSize: fonts.big,
  },
  'text-inline': {
    color: '#FFF',
  },
  'text-outline': {
    color: colors.primary,
  },
});

export default styles;
