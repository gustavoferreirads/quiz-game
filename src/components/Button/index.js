import React, { Fragment } from 'react';
import * as PropTypes from 'prop-types';
import { Text, TouchableOpacity } from 'react-native';
import styles from './styles';

const Button = ({ style, children, type, onPress }) => (
  <TouchableOpacity style={[styles.container, style, styles[`button-${type}`]]} onPress={onPress}>
    <Text style={[styles.text, styles[`text-${type}`]]}>{children}</Text>
  </TouchableOpacity>
);

Button.defaultProps = {
  type: 'inline',
  onPress: () => {},
  style: {},
  children: <Fragment />,
};

Button.propTypes = {
  children: PropTypes.element,
  style: PropTypes.object,
  type: PropTypes.string,
  onPress: PropTypes.func,
};

export default Button;
