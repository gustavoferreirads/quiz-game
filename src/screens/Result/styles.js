import { StyleSheet } from 'react-native';
import { colors, fonts, metrics } from '../../styles';

export const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.lighter,
    flex: 1,
  },
  header: {
    alignItems: 'center',
    flexGrow: 1,
  },
  title: {
    fontSize: fonts.largest,
    color: colors.regular,
    textAlign: 'center',
  },
  graph: {
    marginTop: metrics.mediumSize,
    marginBottom: metrics.mediumSize,
  },
  subTitle: {
    fontSize: fonts.large,
    color: colors.regular,
    textAlign: 'center',
  },
  content: {
    height: 320,
    borderRadius: metrics.radiusMedium,
    marginTop: metrics.size,
    padding: metrics.mediumSize,
    paddingBottom: 0,
    backgroundColor: colors.white,
  },
  actions: {
    padding: metrics.size,
    flexGrow: 1,
    justifyContent: 'space-between',
  },
  item: {
    borderWidth: 1,
    borderRadius: metrics.radius,
    borderStyle: 'solid',
    borderColor: colors.primary,
    marginBottom: metrics.mediumSize,
  },
  description: {
    fontSize: fonts.big,
    color: colors.white,
    textAlign: 'center',
  },
  answer: {
    fontWeight: 'bold',
    marginTop: 10,
  },
  result: {
    flexDirection: 'row',
    padding: metrics.mediumSize,
    justifyContent: 'center',
    alignItems: 'center',
  },
  headerItem: {
    borderTopLeftRadius: metrics.radius,
    borderTopRightRadius: metrics.radius,
    padding: metrics.size,
    paddingBottom: 5,
    backgroundColor: colors.primary,
  },
  text: {
    fontSize: fonts.bigMd,
    marginRight: 10,
  },
  question: {
    margin: 10,
  },
  separator: {
    width: '100%',
    borderColor: colors.light,
    borderEndWidth: 1,
    borderStyle: 'solid',
  },
});

export const progressCircle = {
  radius: 70,
  borderWidth: 10,
  color: colors.primary,
  shadowColor: colors.primaryOpacity,
  bgColor: colors.lighter,
};

export const icons = {
  size: 30,
  like: {
    name: 'like1',
    color: colors.primary,
  },
  unlike: {
    name: 'dislike2',
    color: colors.primaryOpacity,
  },
};
