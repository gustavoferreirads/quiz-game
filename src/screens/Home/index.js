import React from 'react';
import * as PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Text, View } from 'react-native';
import * as actions from '../../redux/actions/quiz';
import Container from '../../components/Container';
import Button from '../../components/Button';

import styles from './styles';

export const Home = ({ navigation, fetchQuestions }) => {
  const navigateBegin = () => {
    fetchQuestions();
    navigation.navigate('Quiz');
  };

  return (
    <Container style={styles.container}>
      <View style={styles.boxTitle}>
        <Text style={styles.title}>Welcome to the</Text>
        <Text style={styles.title}>Trivia Challenge!</Text>
      </View>
      <View style={styles.grow}>
        <Text style={styles.internalText}>
          You will be presented with 10 True or False questions.
        </Text>
      </View>
      <View style={styles.grow}>
        <Text style={[styles.internalText, styles.bolder]}>Can you score 100%?</Text>
      </View>
      <View style={styles.grow}>
        <Button onPress={navigateBegin}>
          <Text>BEGIN</Text>
        </Button>
      </View>
    </Container>
  );
};

Home.defaultProps = {
  navigation: {},
};

Home.propTypes = {
  navigation: PropTypes.object,
  fetchQuestions: PropTypes.func.isRequired,
};

export default connect(
  () => ({}),
  dispatch => bindActionCreators(actions, dispatch),
)(Home);
