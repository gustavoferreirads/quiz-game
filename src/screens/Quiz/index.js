import React from 'react';
import * as PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Text, View } from 'react-native';
import Container from '../../components/Container';
import * as actions from '../../redux/actions/quiz';
import { withLoading } from '../../components/Loading';
import styles from './styles';
import Answer from '../../components/Answer';

const TRUE = 'True';
const FALSE = 'False';

export const Quiz = ({ navigation, current, size, doAnswer }) => {
  const { position, wrapQuestion } = current;

  const onAnswer = answer => {
    doAnswer(answer);
    if (position === size) {
      navigation.navigate('Results');
    }
  };

  return (
    <Container style={styles.container}>
      <Text style={styles.title}>{wrapQuestion.category}</Text>
      <View style={styles.content}>
        <View style={styles.box}>
          <View style={styles.internalBox}>
            <Text style={styles.question}>{wrapQuestion.question}</Text>
          </View>
          <Text style={styles.position}>
            {position} of {size}
          </Text>
        </View>
        <View style={styles.contentAnswer}>
          <Answer option="A" text="True" onPress={() => onAnswer(TRUE)} />
          <Answer option="B" text="False" onPress={() => onAnswer(FALSE)} />
        </View>
      </View>
    </Container>
  );
};

Quiz.defaultProps = {
  current: {},
  navigation: {},
  size: 0,
};

Quiz.propTypes = {
  navigation: PropTypes.object,
  doAnswer: PropTypes.func.isRequired,
  current: PropTypes.object,
  size: PropTypes.number,
};

export default withLoading(
  connect(
    store => ({
      current: store.Quiz.current,
      size: store.Quiz.size,
    }),
    dispatch => bindActionCreators(actions, dispatch),
  )(Quiz),
);
