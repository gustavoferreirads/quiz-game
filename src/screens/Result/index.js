import React, { Fragment } from 'react';
import * as PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { FlatList, Text, View } from 'react-native';
import ProgressCircle from 'react-native-progress-circle';
import { AntDesign } from '@expo/vector-icons/build/Icons';
import Button from '../../components/Button';
import Container from '../../components/Container';
import * as actions from '../../redux/actions/quiz';
import { icons, progressCircle, styles } from './styles';

const { radius, borderWidth, color, shadowColor, bgColor } = progressCircle;

export const Result = ({ navigation, fetchQuestions, answers, score, percent }) => {
  const navigateBegin = () => {
    navigation.navigate('Home');
  };

  const navigateQuiz = () => {
    fetchQuestions();
    navigation.navigate('Quiz');
  };

  return (
    <Container style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.title}>You scored</Text>

        <View style={styles.graph}>
          <ProgressCircle
            percent={percent}
            radius={radius}
            borderWidth={borderWidth}
            color={color}
            shadowColor={shadowColor}
            bgColor={bgColor}
          >
            <Text style={styles.title}>
              {score}/{answers.length}
            </Text>
          </ProgressCircle>
        </View>

        <Text style={styles.subTitle}>You got {percent}% of the answers correct</Text>
        <Text style={styles.subTitle}>Congratulations!</Text>
      </View>

      <View style={styles.content}>
        <FlatList
          data={answers}
          renderItem={({ item }) => (
            <View style={styles.item}>
              <View style={styles.headerItem}>
                <Text style={styles.description}>{item.question}</Text>
                <Text style={[styles.description, styles.answer]}>{item.correct_answer}</Text>
              </View>

              <View style={styles.result}>
                {item.isCorrect && (
                  <Fragment>
                    <Text style={styles.text}>You're right!</Text>
                    <AntDesign name={icons.like.name} size={icons.size} color={icons.like.color} />
                  </Fragment>
                )}

                {!item.isCorrect && (
                  <Fragment>
                    <Text style={styles.text}>You missed!</Text>
                    <AntDesign
                      name={icons.unlike.name}
                      size={icons.size}
                      color={icons.unlike.color}
                    />
                  </Fragment>
                )}
              </View>
            </View>
          )}
          keyExtractor={(item, index) => index.toString()}
        />
      </View>
      <View style={styles.actions}>
        <Button onPress={navigateQuiz}>
          <Text>PLAY AGAIN?</Text>
        </Button>

        <Button type="outline" onPress={navigateBegin}>
          <Text>HOME</Text>
        </Button>
      </View>
    </Container>
  );
};

Result.defaultProps = {
  answers: [],
  navigation: {},
  score: 0,
  percent: 0,
};

Result.propTypes = {
  navigation: PropTypes.object,
  answers: PropTypes.arrayOf(PropTypes.object),
  fetchQuestions: PropTypes.func.isRequired,
  score: PropTypes.number,
  percent: PropTypes.number,
};

export default connect(
  ({ Quiz }) => ({
    answers: Quiz.answers,
    score: Quiz.score,
    percent: Quiz.percent,
  }),
  dispatch => bindActionCreators(actions, dispatch),
)(Result);
