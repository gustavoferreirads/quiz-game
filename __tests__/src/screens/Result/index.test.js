import React from 'react';
import { shallow } from 'enzyme/build';
import { Result } from '../../../../src/screens/Result';
import Button from '../../../../src/components/Button';
import { answers } from './mock';

let fetchQuestions;

const assert = (screen, position) => {
  const navigator = {
    navigate: jest.fn(x => x),
  };

  const wrapper = shallow(
    <Result
      answers={answers}
      navigation={navigator}
      fetchQuestions={fetchQuestions}
      percent={20}
      score={100}
    />,
  );

  wrapper
    .find(Button)
    .at(position)
    .props()
    .onPress();
  expect(navigator.navigate.mock.calls[0][0]).toBe(screen);
  expect(navigator.navigate.mock.calls.length).toBe(1);
};

beforeEach(() => {
  fetchQuestions = jest.fn(() => {});
});

it('renders a Result ', () => {
  expect(
    shallow(
      <Result
        answers={answers}
        navigation={{}}
        fetchQuestions={jest.fn(() => {})}
        percent={20}
        score={100}
      />,
    ),
  ).toMatchSnapshot();
});

it('simulate click on button "Play Again"', () => {
  assert('Quiz', 0);
  expect(fetchQuestions.mock.calls.length).toBe(1);
});

it('simulate click on button "Home"', () => {
  assert('Home', 1);
  expect(fetchQuestions.mock.calls.length).toBe(0);
});
