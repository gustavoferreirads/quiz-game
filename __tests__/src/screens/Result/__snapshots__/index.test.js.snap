// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Result  1`] = `
<Container
  style={
    Object {
      "backgroundColor": "#eee",
      "flex": 1,
    }
  }
>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flexGrow": 1,
      }
    }
  >
    <Component
      style={
        Object {
          "color": "#666",
          "fontSize": 22,
          "textAlign": "center",
        }
      }
    >
      You scored
    </Component>
    <Component
      style={
        Object {
          "marginBottom": 20,
          "marginTop": 20,
        }
      }
    >
      <PercentageCircle
        bgColor="#eee"
        borderWidth={10}
        color="#FF9982"
        containerStyle={null}
        percent={20}
        radius={70}
        shadowColor="rgba(255, 153, 130, 0.3)"
      >
        <Component
          style={
            Object {
              "color": "#666",
              "fontSize": 22,
              "textAlign": "center",
            }
          }
        >
          100
          /
          2
        </Component>
      </PercentageCircle>
    </Component>
    <Component
      style={
        Object {
          "color": "#666",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      You got 
      20
      % of the answers correct
    </Component>
    <Component
      style={
        Object {
          "color": "#666",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      Congratulations!
    </Component>
  </Component>
  <Component
    style={
      Object {
        "backgroundColor": "#FFF",
        "borderRadius": 20,
        "height": 320,
        "marginTop": 10,
        "padding": 20,
        "paddingBottom": 0,
      }
    }
  >
    <FlatList
      data={
        Array [
          Object {
            "category": "Entertainment: Film",
            "correct_answer": "False",
            "difficulty": "hard",
            "incorrect_answers": Array [
              "True",
            ],
            "isCorrect": true,
            "question": "&quot;Cube&quot;, &quot;Cube 2: Hypercube&quot; and &quot;Cube Zero&quot; were directed by the same person.",
            "type": "boolean",
          },
          Object {
            "category": "General Knowledge",
            "correct_answer": "True",
            "difficulty": "hard",
            "incorrect_answers": Array [
              "False",
            ],
            "question": "In Scandinavian languages, the letter &Aring; means river.",
            "type": "boolean",
          },
        ]
      }
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </Component>
  <Component
    style={
      Object {
        "flexGrow": 1,
        "justifyContent": "space-between",
        "padding": 10,
      }
    }
  >
    <Button
      onPress={[Function]}
      style={Object {}}
      type="inline"
    >
      <Component>
        PLAY AGAIN?
      </Component>
    </Button>
    <Button
      onPress={[Function]}
      style={Object {}}
      type="outline"
    >
      <Component>
        HOME
      </Component>
    </Button>
  </Component>
</Container>
`;
